### PR TITLE
Training NLVR parser using logical forms

### PR DESCRIPTION
**Pull request by pdasigi**
Summary of changes:
1) Use multi-world consistency along with agenda in `scripts/nlvr/get_nlvr_logical_forms.py` and added a flag `--ignore_agenda` to the script to produce logical forms relying only on consistency (they're bad!).
2) Changed the format of the preprocessed files produced by `scripts/nlvr/get_nlvr_logical_forms.py` to include action sequences instead of logical forms, and also worlds and labels in the output for computing metrics during training.
3) Changed the dataset reader to allow reading preprocessed data format.
4) Kept the core functionality of `NlvrSemanticParser` in a base class, and broke the rest into two classes - `NlvrCoverageSemanticParser`, that trains without logical form supervision using ERM as before; and `NlvrDirectSemanticParser` (for the lack of a better name), that trains with the logical forms produced by offline search using MML.

### CODE REVIEW

File: allennlp/models/encoder_decoders/nlvr/nlvr_direct_semantic_parser.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+    def forward(self,  # type: ignore
+                sentence: Dict[str, torch.LongTensor],
+                worlds: List[List[NlvrWorld]],
+                actions: List[List[ProductionRuleArray]],
+                target_action_sequences: torch.LongTensor = None,
+                labels: torch.LongTensor = None) -> Dict[str, torch.Tensor]:
+        # pylint: disable=arguments-differ
+        """
+        Decoder logic for producing type constrained target sequences that maximize coverage of
+        their respective agendas, and minimize a denotation based loss.
```
Comment (matt-gardner) [inline allennlp/models/encoder_decoders/nlvr/nlvr_direct_semantic_parser.py]: This docstring is out of date.

File: allennlp/models/encoder_decoders/nlvr/nlvr_coverage_semantic_parser.py [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+        target_checklist = nn_util.new_variable_with_data(agenda,
+                                                          torch.Tensor(target_checklist_list))
+        if self._penalize_non_agenda_actions:
+            # All terminal actions are relevant
+            checklist_mask = torch.ones_like(target_checklist)
+        else:
+            checklist_mask = (target_checklist != 0).float()
+        return target_checklist, terminal_actions, checklist_mask
+
+    def _update_metrics(self,
```
Comment (matt-gardner) [inline allennlp/models/encoder_decoders/nlvr/nlvr_coverage_semantic_parser.py]: Maybe possible to share this code between the two classes, but probably not worth worrying about at this point.  Once we're to the cleaning-up-the-code stage, we can think about this more.

File: allennlp/data/dataset_readers/nlvr.py [hunk 3]
```diff
@@ -107,13 +107,28 @@ def _read(self, file_path: str):
                     # one element for consistency.
                     labels = [data["label"]]
                     structured_representations = [data["structured_rep"]]
-                yield self.text_to_instance(sentence, structured_representations, labels)
+
+                target_sequences: List[List[str]] = None
+                # TODO(pradeep): The processed file also has incorrect sequences as well, which are
+                # needed if we want to define some sort of a hinge-loss based trainer. Deal with
+                # them.
+                if "correct_sequences" in data:
```
[the remark this answers is not in the document]
Comment (pdasigi) [reply allennlp/data/dataset_readers/nlvr.py]: Done. Thanks!

File: allennlp/models/encoder_decoders/nlvr/nlvr_decoder_state.py [hunk 4]
```diff
@@ -29,24 +29,6 @@ class NlvrDecoderState(DecoderState['NlvrDecoderState']):
     grammar_state : ``List[GrammarState]``
         This hold the current grammar state for each element of the group.  The ``GrammarState``
         keeps track of which actions are currently valid.
-    terminal_actions : ``List[torch.Tensor]``
```
[the remark this answers is not in the document]
Comment (pdasigi) [reply allennlp/models/encoder_decoders/nlvr/nlvr_decoder_state.py]: Agreed. Added a TODO.

File: allennlp/semparse/action_space_walker.py [hunk 5]
```diff
@@ -102,3 +103,15 @@ def get_logical_forms_with_agenda(self,
...
         logical_forms = [self._world.get_logical_form(path) for path in paths]
         return logical_forms
+
+    def get_all_logical_forms(self,
+                              max_num_logical_forms: int = None) -> List[str]:
+        if self._completed_paths is None:
+            self._walk()
+        paths = self._completed_paths
+        if max_num_logical_forms is not None:
+            if self._length_sorted_paths is None:
```
[the remark this answers is not in the document]
Comment (pdasigi) [reply allennlp/semparse/action_space_walker.py]: That's correct.

File: scripts/nlvr/get_nlvr_logical_forms.py [hunk 6]
```diff
@@ -43,30 +45,55 @@ def process_data(input_file: str,
...
+                                   "worlds": structured_reps,
+                                   "labels": label_strings})
+        else:
+            processed_data.append({"id": instance_id,
+                                   "sentence": sentence,
+                                   "correct_logical_forms": correct_logical_forms,
+                                   "incorrect_logical_forms": incorrect_logical_forms,
+                                   "worlds": structured_reps,
+                                   "labels": label_strings})
     outfile = open(output_file, "w")
```
[the remark this answers is not in the document]
Comment (pdasigi) [reply scripts/nlvr/get_nlvr_logical_forms.py]: Done. Thanks!

File: scripts/nlvr/get_nlvr_logical_forms.py [hunk 7]
```diff
@@ -78,5 +105,13 @@ def process_data(input_file: str,
     parser.add_argument("--max_num_logical_forms", type=int,
                         help="Maximum number of logical forms per denotation, per question",
                         default=20)
+    parser.add_argument("--ignore_agenda", dest="ignore_agenda", help="Should we ignore the "
```
[the remark this answers is not in the document]
Comment (pdasigi) [reply scripts/nlvr/get_nlvr_logical_forms.py]: Done.

File: scripts/nlvr/get_nlvr_logical_forms.py [hunk 8]
```diff
@@ -78,5 +105,13 @@ def process_data(input_file: str,
...
+                        "agenda and use consistency as the only signal to get logical forms?",
+                        action='store_true')
+    parser.add_argument("--write_action_sequences", dest="write_sequences", help="If this "
+                        "flag is set, action sequences instead of logical forms will be written "
+                        "to the json file. This will avoid having to parse the logical forms again "
+                        "in the NlvrDatasetReader.", action='store_true')
     args = parser.parse_args()
-    process_data(args.input, args.output, args.max_path_length, args.max_num_logical_forms)
+    process_data(args.input, args.output, args.max_path_length,
```
[the remark this answers is not in the document]
Comment (pdasigi) [reply scripts/nlvr/get_nlvr_logical_forms.py]: Done.

File: tests/data/dataset_readers/nlvr_test.py [hunk 9]
```diff
@@ -5,6 +5,7 @@
 
 
 class TestNlvrDatasetReader(AllenNlpTestCase):
+    # TODO(pradeep): Write a test for reading grouped and processed data.
```
[the remark this answers is not in the document]
Comment (pdasigi) [reply tests/data/dataset_readers/nlvr_test.py]: I forgot what that comment refers to. It's probably out of date, because the test for processed data works with both grouped and ungrouped files. Removed it.

File: allennlp/models/encoder_decoders/nlvr/nlvr_direct_semantic_parser.py (hunk 1 above)
Comment (pdasigi) [reply allennlp/models/encoder_decoders/nlvr/nlvr_direct_semantic_parser.py]: Thanks! Fixed it.

File: allennlp/models/encoder_decoders/nlvr/nlvr_coverage_semantic_parser.py (hunk 2 above)
Comment (pdasigi) [reply allennlp/models/encoder_decoders/nlvr/nlvr_coverage_semantic_parser.py]: Right. Added a TODO.
